server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -388,6 +390,12 @@ private static boolean upgradeLegacyNodeFolders(Logger logger, Settings settings
             throw new IllegalStateException(message, e);
         }
 
+        try {
+            checkForIndexCompatibility(logger, legacyNodeLock.getNodePaths());
+        } finally {
+            legacyNodeLock.close();
```

Comment:
I think we should only release the `legacyNodeLock` on failure - on success we need to keep hold of it until the end of this method.

We could move the `checkForIndexCompatibility` call within the next `try` block right?

Reply:
Ah good point, I was fixing a bug and you are right, I release it too soon.